R/appMetadata.R

```diff
@@ -0,0 +1,378 @@
+    return("shiny")
+  }
+
+  # Determine if we have Rmd files, and if they use the Shiny runtime.
+  rmdFiles <- grep("^[^/\\\\]+\\.rmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
+  shinyRmdFiles <- sapply(file.path(appDir, rmdFiles), isShinyRmd)
+
+  # Determine if we have qmd files, and if they use the Shiny runtime
+  qmdFiles <- grep("^[^/\\\\]+\\.qmd$", appFiles, ignore.case = TRUE, perl = TRUE, value = TRUE)
+  shinyQmdFiles <- sapply(file.path(appDir, qmdFiles), isShinyRmd)
+
+  # We make Quarto requirement conditional on the presence of files that Quarto
+  # can render and _quarto.yml, because keying off the presence of qmds
+  # *or* _quarto.yml was causing deployment failures in static content.
+  # https://github.com/rstudio/rstudio/issues/11444
+  hasQuartoYaml <- any(grepl("^_quarto.y(a)?ml$", x = appFiles, ignore.case = TRUE, perl = TRUE))
+  hasQuartoCompatibleFiles <- length(qmdFiles) > 0 || length(rmdFiles > 0)
+  requiresQuarto <- (hasQuartoCompatibleFiles && hasQuartoYaml) || length(qmdFiles) > 0
+
+  # We gate the deployment of content that appears to be Quarto behind the
+  # presence of Quarto metadata. Rmd files can still be deployed as Quarto
+  # content.
+  # TODO(HW): I think this logic should be moved into appMetdata() because this
+  # should also error if the user specified appMode = "quarto-shiny"
+  if (requiresQuarto && is.null(quartoInfo)) {
+    stop(paste(
+      "Attempting to deploy Quarto content without Quarto metadata.",
+      "Please provide the path to a quarto binary to the 'quarto' argument."
+    ))
+  }
+
+  # Shiny or Quarto documents with "server: shiny" in their YAML front matter
+  # are rmd-shiny or quarto-shiny.
+  if (any(shinyRmdFiles) || any(shinyQmdFiles)) {
+    if (!is.null(quartoInfo)) {
+      return("quarto-shiny")
+    } else {
+      return("rmd-shiny")
+    }
+  }
+
+  # Shiny application using server.R; checked later than Rmd with shiny runtime
+  # because server.R may contain the server code paired with a ShinyRmd and needs
+  # to be run by rmarkdown::run (rmd-shiny).
+  serverR <- grep("^server.r$", appFiles, ignore.case = TRUE, perl = TRUE)
+  if (length(serverR) > 0) {
+    return("shiny")
+  }
+
+  # Any non-Shiny R Markdown or Quarto documents are rendered content and get
+  # rmd-static or quarto-static.
+  if (length(rmdFiles) > 0 || length(qmdFiles) > 0) {
+    if (!is.null(quartoInfo)) {
+      return("quarto-static")
+    } else {
+      # For Shinyapps and posit.cloud, treat "rmd-static" app mode as "rmd-shiny" so that
+      # they can be served from a shiny process in Connect until we have better support of
+      # rmarkdown static content
+      if (isCloudServer) {
+        return("rmd-shiny")
+      }
+      return("rmd-static")
+    }
+  }
+
+  # We don't have an RMarkdown, Shiny app, or Plumber API, but we have a saved model
+  if (length(grep("(saved_model.pb|saved_model.pbtxt)$", appFiles, ignore.case = TRUE, perl = TRUE)) > 0) {
```

Comment:
FYI: I have outstanding work to remove support for the TF app-mode from `rsconnect`. We're letting the most recent Connect release (2023.01.1) age a little before making that change.

Reply:
Good to know, because this regular expression is inconsistent with the others.